# Hands-On Assignment 3

In this assignment, you will learn the basics of data cleaning and munging (manipulation).
Data is the heart and soul of a data scientist and machine learner.
Therefore, it is important for you to get comfortable taking data from its natural, messy state
to a state where it is ready for machine learning algorithms.

Since data is at the core of all we do, we learn to love data and treat it well.
However, not everyone is a data scientist, a computer scientist, or even a programmer.
Others that you work with may carefully collect very important data,
but handling and storing the data may not be their expertise.
So whenever working with data prepared by others,
we must always treat their data with care and remember that even data presented in suboptimal ways can be of the utmost importance.
Also remember that even as we learn better data handling practices,
there will still be those with data handling skill and methodologies that are leagues better than our own.
(It's database people, they are usually the best data people. (Written by Eriq, a database person.))

The objective of this assignment is for you to learn about:
 - Finding errors and inconsistencies in data.
 - Extracting key pieces of information from a data column.
 - Assigning types to data columns.
 - Understanding and finding outliers in your data.
 - Replacing missing pieces of data.
 - Encoding data.
 - Joining data.

## Data: CIA World Factbook - 2022

For this assignment, we will be using data from the [CIA World Factbook](https://www.cia.gov/the-world-factbook/).
This is a collection of real data about different countries and regions in the world compiled every year by the CIA.
Specifically, we will be using the world factbook data from [this repository](https://github.com/factbook/factbook.json/) from 2022.

The data can be found in your assignment repository in the `cia_world_factbook_2022.json` file.

Let's take a look at the data as we have done in previous assignments:

In [29]:
import json
import re
import pandas
import numpy as np

# Our data comes in JSON, so first parse the JSON data.
with open('cia_world_factbook_2022.json', 'r') as file:
    data = json.load(file)

# Now load the JSON data into a Pandas dataframe.
world_data = pandas.DataFrame.from_dict(data, orient = 'index')

# Sort the data by country name so it looks nice.
world_data.sort_index(axis = 0, inplace = True)

# Move the country name out of the index and into an actual column.
world_data.insert(0, 'Country', world_data.index)
world_data.reset_index(drop = True, inplace = True)

world_data

Country Life expectancy   
0       Afghanistan     53.65 years  \
1          Akrotiri             NaN   
2           Albania     79.47 years   
3           Algeria     78.03 years   
4    American Samoa     75.32 years   
..              ...             ...   
245  Western Sahara      64.5 years   
246           Yemen     67.51 years   
247          Zambia     66.26 years   
248        Zimbabwe     63.32 years   
249            none     74.44 years   

                                            Population Literacy   
0                               38,346,720 (2022 est.)    37.3%  \
1    (2020) approximately 18,195 on the Sovereign B...      NaN   
2                                3,095,344 (2022 est.)    98.1%   
3                               44,178,884 (2022 est.)    81.4%   
4                                   45,443 (2022 est.)       NA   
..                                                 ...      ...   
245                           652,271 (July 2020 est.)      NaN   
246                             30,984,689 (2022 est.)    70.1%   
247                             19,642,123 (2022 est.)    86.7%   
248                             15,121,004 (2022 est.)    86.5%   
249                                101,009 (2022 est.)       NA   

    Electricity access  Internet access         Improved Sanitation   
0           99% (2018)  18% (2020 est.)  total: 61.4% of population  \
1                  NaN              NaN                         NaN   
2          100% (2020)  72% (2020 est.)  total: 99.9% of population   
3         99.4% (2019)  63% (2020 est.)  total: 96.5% of population   
4                  NaN  31% (2020 est.)    total: 99% of population   
..                 ...              ...                         ...   
245                NaN              NaN                         NaN   
246         47% (2019)  27% (2019 est.)  total: 59.1% of population   
247         37% (2019)  20% (2020 est.)  total: 51.7% of population   
248         53% (2019)  29% (2020 est.)  total: 64.2% of population   
249         82% (2018)  35% (2019 est.)  total: 88.3% of population   

                       Improved Water   
0    total: 76.5% of population 70.2%  \
1                           total: NA   
2            total: 97% of population   
3          total: 99.4% of population   
4          total: 99.8% of population   
..                                ...   
245                               NaN   
246        total: 99.6% of population   
247        total: 71.6% of population   
248        total: 76.9% of population   
249        total: 78.6% of population   

                                    Export commodities Labor Agriculture  ...   
0    gold, grapes, opium, fruits and nuts, insect r...             44.3%  ...  \
1                                                  NaN               NaN  ...   
2    leather footwear and parts, crude petroleum, i...             41.4%  ...   
3    crude petroleum, natural gas, refined petroleu...             10.8%  ...   
4                                          canned tuna                NA  ...   
..                                                 ...               ...  ...   
245                         phosphates 62% (2012 est.)               50%  ...   
246  crude petroleum, gold, fish, industrial chemic...               NaN  ...   
247  copper, gold, gemstones, sulfuric acid, raw su...             54.8%  ...   
248  gold, tobacco, iron alloys, nickel, diamonds, ...             67.5%  ...   
249  fish and fish products, coral/shells, scrap me...              0.9%  ...   

    Unemployment rate 2012 Unemployment rate 2014 Unemployment rate 1998   
0                      NaN                    NaN                    NaN  \
1                      NaN                    NaN                    NaN   
2                      NaN                    NaN                    NaN   
3                      NaN                    NaN                    NaN   
4                      NaN             

Let's also take a look at the column information:

In [30]:
world_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 35 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   Country                       250 non-null    object
 1   Life expectancy               233 non-null    object
 2   Population                    250 non-null    object
 3   Literacy                      229 non-null    object
 4   Electricity access            219 non-null    object
 5   Internet access               227 non-null    object
 6   Improved Sanitation           230 non-null    object
 7   Improved Water                247 non-null    object
 8   Export commodities            232 non-null    object
 9   Labor Agriculture             212 non-null    object
 10  Labor Industry                207 non-null    object
 11  Labor Services                198 non-null    object
 12  Unemployment rate 2021        187 non-null    object
 13  Unemployment rate 20

Now the numerical values:

In [31]:
world_data.describe()

Country Life expectancy   
count           250             233  \
unique          250             207   
top     Afghanistan              NA   
freq              1               7   

                                        Population Literacy   
count                                          250      229  \
unique                                         243      118   
top     (July 2021 est.) no indigenous inhabitants       NA   
freq                                             4       66   

       Electricity access  Internet access        Improved Sanitation   
count                 219              227                        230  \
unique                 68              132                        122   
top           100% (2020)  78% (2020 est.)  total: 100% of population   
freq                  125               10                         54   

                   Improved Water   
count                         247  \
unique                        102   
top     total: 100% of population   
freq                           54   

                                       Export commodities Labor Agriculture   
count                                                 232               212  \
unique                                                232               154   
top     gold, grapes, opium, fruits and nuts, insect r...                NA   
freq                                                    1                 7   

        ... Unemployment rate 2012 Unemployment rate 2014   
count   ...                      3                      3  \
unique  ...                      3                      3   
top     ...              2% (2012)        11% (2014 est.)   
freq    ...                      1                      1   

       Unemployment rate 1998 Unemployment rate 2015 Unemployment rate 2002   
count                       1                      7                      1  \
unique                      1                      7                      1   
top            6% (1998 est.)       4.1% (2015 est.)              8% (2002)   
freq                        1                      1                      1   

       Unemployment rate 2018 Unemployment rate April 2011   
count                       1                            1  \
unique                      1                            1   
top         3.69% (2018 est.)         2% (April 2011 est.)   
freq                        1                            1   

       Unemployment rate 2006 Unemployment rate 2005 Unemployment rate 2001  
count                       1                      4                      1  
unique                      1                      4                      1  
top           36% (2006 est.)           29.8% (2005)             12% (2001)  
freq                        1                      1                      1  

[4 rows x 35 columns]

Note that when we tried to get information about the numerical columns like in a previous assignment,
we didn't get nearly as much information as we did before.
If we look back to `world_data.info()`,
we can see this is because Pandas doesn't actually know which of our columns are numbers
(the dtype ("data type") for each column is just `object`).
If we look at some of the values (like "38,346,720 (2022 est.)"),
we can see why Pandas would be confused about how to treat that data.
Pandas will always try and choose a data type that can be applied to every value in the column.
So if a column has a million ints and one float, then the column type will be float
(since all ints are floats, but not all floats are ints).

Let's see if we can cleanup this data!

## Part 1: Exploring Columns

The first thing to do with a new dataset is explore the columns.
 - What data is each column suppose to contain?
 - What data type should be used to represent each column?
 - Which columns may be useful?

Not all columns in each dataset are going to be useful.
We can start our work by eliminating columns that have a low chance of being needed.

<h3 style="color: darkorange";>★ Task 1.A</h3>

Complete the following function that takes in a frame and drops the columns in the frame that have greater than `sparsity_threshold` percent of values empty.
Assume any value that [Pandas.isna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isna.html) returns true for is "empty".
Return the passed-in frame.

In [32]:
def drop_sparse_columns(frame, sparsity_threshold):
    return frame.dropna(axis = 1, thresh = sparsity_threshold * len(frame))

print("Dropping any column with at least 50% empty values:")
drop_sparse_columns(world_data, 0.50)

Dropping any column with at least 50% empty values:


Country Life expectancy   
0       Afghanistan     53.65 years  \
1          Akrotiri             NaN   
2           Albania     79.47 years   
3           Algeria     78.03 years   
4    American Samoa     75.32 years   
..              ...             ...   
245  Western Sahara      64.5 years   
246           Yemen     67.51 years   
247          Zambia     66.26 years   
248        Zimbabwe     63.32 years   
249            none     74.44 years   

                                            Population Literacy   
0                               38,346,720 (2022 est.)    37.3%  \
1    (2020) approximately 18,195 on the Sovereign B...      NaN   
2                                3,095,344 (2022 est.)    98.1%   
3                               44,178,884 (2022 est.)    81.4%   
4                                   45,443 (2022 est.)       NA   
..                                                 ...      ...   
245                           652,271 (July 2020 est.)      NaN   
246                             30,984,689 (2022 est.)    70.1%   
247                             19,642,123 (2022 est.)    86.7%   
248                             15,121,004 (2022 est.)    86.5%   
249                                101,009 (2022 est.)       NA   

    Electricity access  Internet access         Improved Sanitation   
0           99% (2018)  18% (2020 est.)  total: 61.4% of population  \
1                  NaN              NaN                         NaN   
2          100% (2020)  72% (2020 est.)  total: 99.9% of population   
3         99.4% (2019)  63% (2020 est.)  total: 96.5% of population   
4                  NaN  31% (2020 est.)    total: 99% of population   
..                 ...              ...                         ...   
245                NaN              NaN                         NaN   
246         47% (2019)  27% (2019 est.)  total: 59.1% of population   
247         37% (2019)  20% (2020 est.)  total: 51.7% of population   
248         53% (2019)  29% (2020 est.)  total: 64.2% of population   
249         82% (2018)  35% (2019 est.)  total: 88.3% of population   

                       Improved Water   
0    total: 76.5% of population 70.2%  \
1                           total: NA   
2            total: 97% of population   
3          total: 99.4% of population   
4          total: 99.8% of population   
..                                ...   
245                               NaN   
246        total: 99.6% of population   
247        total: 71.6% of population   
248        total: 76.9% of population   
249        total: 78.6% of population   

                                    Export commodities Labor Agriculture   
0    gold, grapes, opium, fruits and nuts, insect r...             44.3%  \
1                                                  NaN               NaN   
2    leather footwear and parts, crude petroleum, i...             41.4%   
3    crude petroleum, natural gas, refined petroleu...             10.8%   
4                                          canned tuna                NA   
..                                                 ...               ...   
245                         phosphates 62% (2012 est.)               50%   
246  crude petroleum, gold, fish, industrial chemic...               NaN   
247  copper, gold, gemstones, sulfuric acid, raw su...             54.8%   
248  gold, tobacco, iron alloys, nickel, diamonds, ...             67.5%   
249  fish and fish products, coral/shells, scrap me...              0.9%   

    Labor Industry     Labor Services Unemployment rate 2021   
0            18.1%  37.6% (2017 est.)     13.28% (2021 est.)  \
1              NaN                NaN                    NaN   
2            18.3%  40.3% (2017 est.)     11.82% (2021 est.)   
3            30.9%  58.4% (2011 est.)      12.7% (2021 est.)   
4            15.5%  46.4% (2015 est.)                    NaN   
..             ...                ...                    ...   
245            50%             

Now that we have fewer columns to worry about,
we can take a closer look at our data.

It looks like this data was originally meant for people and not machines.
There are several things that make it easier for people to read, but harder for us machine learning folks:
 - Commas in numeric values: '38,346,720' instead of '38346720'.
 - Units after numbers: '53.65 years', '37.3%'.
 - Notes about the data interweaved with the data itself: '37.6% (2017 est.)' instead of just '37.6%' (or better: '0.376').

Externally displaying data with these traits can really help analysts and non-data scientists,
but can really get in the way of our data analysis work.
When possible, we want to store the data cleanly and translate it to a more human readable form (often called a *report*) for readers, analysts, managers, etc..
This also makes it easier to transform our data in different ways
(like using a dot instead of a comma as a decimal separator in some countries: '1,000' vs '1.000').

In our data, it looks like many columns just contain a single main number with some potential context information around it:

In [33]:
world_data['Population'][0]

'38,346,720 (2022 est.)'

There are also more complex cases that contain multiple relevant numbers:

In [34]:
world_data['Population'][1]

'(2020) approximately 18,195 on the Sovereign Base Areas of Akrotiri and Dhekelia including 11,000 Cypriots and 7,195 Service and UK-based contract personnel and dependents'

We will want to clean up these cells so that they contain just the core piece of information and not all that extra information.

### An Aside on Regular Expressions

[Regular expressions](https://en.wikipedia.org/wiki/Regular_expression) (also called "regexp" or "regex") are patterns that let you find matching text.
Think of them like mathematical expressions for text (an equation can define a line (collection of points), and a regex can define a set of strings).
Covering regular expressions is outside the scope of this course,
and they are **not necessary for any of the assignments in this course**.
However, they can be very useful in many tasks (especially in data science).
And once you get comfortable using them, you start to see how they can be used in almost all of your everyday coding.

<center><img src="xkcd-regular-expressions.png"/></center>
<center style='font-size: small'>Comic courtesy of <a href='https://xkcd.com/208'>xkcd</a></center>

We encourage you to learn about regular expressions and become comfortable using them in your everyday coding.
Here are some resources to help you:
 - [Text Tutorial](https://www.sitepoint.com/learn-regex/)
 - [Video Tutorial](https://www.youtube.com/watch?v=sa-TUpSx1JA)
 - [Cheat Sheet](https://cheatography.com/davechild/cheat-sheets/regular-expressions/)
 - [Regex Playground](https://regex101.com/) (Interactivley create, test, and visualize regular expressions.)

<h3 style="color: darkorange";>★ Task 1.B</h3>

Complete the following function that takes in a frame and a list of columns to ignore.
For all columns that are not ignored,
extract a number from the cell's text and replace the cell's contents with that correctly typed number (either a `float` or `int`).
The number should be converted to an `int` if the text does not have a decimal point (and is not a percentage).
If the cell contains no number, replace it with a `numpy.nan`.
Convert percentages to normalized floats, e.g., `str('37.6% (2017 est.)')` turns into `float(0.376)`.
Return the passed-in frame.

For cells with ambiguous data/numbers,
it is your job to do your best to honor what that data's author likely intended.

**Warning:
This is a hard task with no exact answer (as with most tasks in machine learning).
Look through the data to find as many edge cases as you can, turn those into test cases, and see how many you can solve.
Start with a simple solution and work up from there.**

In [39]:
def extract_numbers(frame, ignore_columns = []):
    for col in frame.columns:
        if col not in ignore_columns:
            for i, cell in enumerate(frame[col]):
                # check if cell is a percentage
                if "%" in str(cell):
                    # extract the percentage as a float and divide by 100 to normalize
                    cell_value = re.findall(r"[\d\.]+", str(cell))[0]
                    cell_value = float(cell_value) / 100.0
                else:
                    # extract the number as a float or int
                    cell_value = re.findall(r"\d+(?:,\d+)*(?:\.\d+)?", str(cell))
                    if len(cell_value) == 0:
                        # if there is no number, replace with NaN
                        cell_value = np.nan
                    else:
                        # if there is a number, convert to float or int
                        cell_value = float(cell_value[0].replace(",", ""))
                        if cell_value.is_integer():
                            cell_value = int(cell_value)
                # update the cell with the cleaned value
                frame.iloc[i, frame.columns.get_loc(col)] = cell_value
    return frame

print("Cleaning up the numeric values:")
extract_numbers(world_data, ['Country', 'Export commodities'])

Cleaning up the numeric values:


Country Life expectancy Population Literacy Electricity access   
0       Afghanistan             NaN        NaN    0.373               0.99  \
1          Akrotiri             NaN        NaN      NaN                NaN   
2           Albania             NaN        NaN    0.981                  1   
3           Algeria             NaN        NaN    0.814              0.994   
4    American Samoa             NaN        NaN      NaN                NaN   
..              ...             ...        ...      ...                ...   
245  Western Sahara             NaN        NaN      NaN                NaN   
246           Yemen             NaN        NaN    0.701               0.47   
247          Zambia             NaN        NaN    0.867               0.37   
248        Zimbabwe             NaN        NaN    0.865               0.53   
249            none             NaN        NaN      NaN               0.82   

    Internet access Improved Sanitation Improved Water   
0              0.18               0.614          0.765  \
1               NaN                 NaN            NaN   
2              0.72               0.999           0.97   
3              0.63               0.965          0.994   
4              0.31                0.99          0.998   
..              ...                 ...            ...   
245             NaN                 NaN            NaN   
246            0.27               0.591          0.996   
247             0.2               0.517          0.716   
248            0.29               0.642          0.769   
249            0.35               0.883          0.786   

                                    Export commodities Labor Agriculture  ...   
0    gold, grapes, opium, fruits and nuts, insect r...             0.443  ...  \
1                                                  NaN               NaN  ...   
2    leather footwear and parts, crude petroleum, i...             0.414  ...   
3    crude petroleum, natural gas, refined petroleu...             0.108  ...   
4                                          canned tuna               NaN  ...   
..                                                 ...               ...  ...   
245                         phosphates 62% (2012 est.)               0.5  ...   
246  crude petroleum, gold, fish, industrial chemic...               NaN  ...   
247  copper, gold, gemstones, sulfuric acid, raw su...             0.548  ...   
248  gold, tobacco, iron alloys, nickel, diamonds, ...             0.675  ...   
249  fish and fish products, coral/shells, scrap me...             0.009  ...   

    Unemployment rate 2012 Unemployment rate 2014 Unemployment rate 1998   
0                      NaN                    NaN                    NaN  \
1                      NaN                    NaN                    NaN   
2                      NaN                    NaN                    NaN   
3                      NaN                    NaN                    NaN   
4                      NaN                    NaN                    NaN   
..                     ...                    ...                    ...   
245                    NaN                    NaN                    NaN   
246                    NaN                    NaN                    NaN   
247                    NaN                    NaN                    NaN   
248                    NaN                    NaN                    NaN   
249                    NaN                    NaN                    NaN   

    Unemployment rate 2015 Unemployment rate 2002 Unemployment rate 2018   
0                      NaN                    NaN                    NaN  \
1                      NaN                    NaN                    NaN   
2                      NaN                    NaN                    NaN   
3                      NaN                    NaN                    NaN   
4                      NaN                    NaN                    NaN   
..                     ...          

As a consequence of what we just did, we also standardized all the NaN values in those columns.
Before we could see empty values represented in several different ways:
 - 'NaN'
 - 'NA'
 - 'total: NA'
 - '' (empty)

Now, since any cell that didn't contain a number was replaced with `numpy.nan`,
all numeric cells with missing/empty values are consistent.

Now that we have cleaned up our numeric columns,
let's officially tell Pandas the data types for our columns.
Because even though most of our columns now only contains numbers (and `numpy.nan`),
Pandas still needs to be told what data type each column uses.

In [ ]:
world_data.describe()

<h3 style="color: darkorange";>★ Task 1.C</h3>

Complete the following function that takes in a frame and assigns the closest matching type to each column.
You can choose from the following three types: `int`, `float`, and `str`.
(Note that columns containing only integers should have the `int` type, not the `float` type.)
Ignore any `numpy.nan` values when making decisions.
Return the passed-in frame.

In [8]:
def guess_types(frame):
    return frame.apply(lambda x: pandas.to_numeric(x, errors = 'ignore'))

print("Data description after types are assigned:")
guess_types(world_data)
world_data.describe()

Data description after types are assigned:


Country Life expectancy   
count           250             233  \
unique          250             207   
top     Afghanistan              NA   
freq              1               7   

                                        Population Literacy   
count                                          250      229  \
unique                                         243      118   
top     (July 2021 est.) no indigenous inhabitants       NA   
freq                                             4       66   

       Electricity access  Internet access        Improved Sanitation   
count                 219              227                        230  \
unique                 68              132                        122   
top           100% (2020)  78% (2020 est.)  total: 100% of population   
freq                  125               10                         54   

                   Improved Water   
count                         247  \
unique                        102   
top     total: 100% of population   
freq                           54   

                                       Export commodities Labor Agriculture   
count                                                 232               212  \
unique                                                232               154   
top     gold, grapes, opium, fruits and nuts, insect r...                NA   
freq                                                    1                 7   

        ... Unemployment rate 2012 Unemployment rate 2014   
count   ...                      3                      3  \
unique  ...                      3                      3   
top     ...              2% (2012)        11% (2014 est.)   
freq    ...                      1                      1   

       Unemployment rate 1998 Unemployment rate 2015 Unemployment rate 2002   
count                       1                      7                      1  \
unique                      1                      7                      1   
top            6% (1998 est.)       4.1% (2015 est.)              8% (2002)   
freq                        1                      1                      1   

       Unemployment rate 2018 Unemployment rate April 2011   
count                       1                            1  \
unique                      1                            1   
top         3.69% (2018 est.)         2% (April 2011 est.)   
freq                        1                            1   

       Unemployment rate 2006 Unemployment rate 2005 Unemployment rate 2001  
count                       1                      4                      1  
unique                      1                      4                      1  
top           36% (2006 est.)           29.8% (2005)             12% (2001)  
freq                        1                      1                      1  

[4 rows x 35 columns]

Now we can see more complete statistics for our numerical columns.
Our data is so much cleaner than when we started,
but it still has some issues lurking in it and is not ready for our machine learning algorithms.

## Part 2: Data Cleaning and Smoothing

Now that our data is cleaner, we can look closer at our numbers.
It is always important to take time to go over your data.
To help understand our data, we will use both aggregate statistics and manual inspection.

### Data Outliers

Now, you may notice some inconsistencies (depending on how well you did in previous parts)
in your data statistics (`describe()`).
For example, you may notice a value for a percentage column above 1.0 or a population in the single digits.
Looking at these aggregate statistics can help us recognize that there are issues in our data.
However, we still need to isolate the specific entries with issues and fix them.

To start, we want to look for [outliers](https://en.wikipedia.org/wiki/Outlier) in our data.
An outlier is just a point that lies "outside" the expected range for our data.
The exact definition for what is considered "outside" our expected range depends on the dataset and situation.

<h3 style="color: darkorange";>★ Task 2.A</h3>

Complete the function below which takes in a frame, a number of standard deviations, and the name of a label column.
This function will search all numeric columns for outliers.
An outlier will be any value that is more that the specified number of standard deviations away from the column's mean.
Any NaNs should be ignored.

The return value should be a dict where the keys are the names of columns that contain an outlier
and the value is a list of tuples with the value of the label column for that row and the outlier value.
Columns without outliers (and non-numeric columns) should be omitted from the returned result.

For example, a valid return from `find_outliers(world_data, 4.0, 'Country')` could be:
```
{
    'Population': [
        ('China', 1410539758.0),
        ('India', 1389637446.0)
    ]
}
```

Use this function as a chance to catch bugs you may have in your number parsing function.

In [97]:
def find_outliers(frame, deviations, label_column_name):
    # initialize an empty dictionary to store outlier information
    outlier_info = {}
    
    # loop over all numeric columns
    for col in frame.select_dtypes(include=np.number).columns:
        # calculate the mean and standard deviation of the column
        col_mean = frame[col].mean()
        col_std = frame[col].std()
        
        # identify values that are more than "deviations" standard deviations away from the mean
        lower_bound = col_mean - deviations * col_std
        upper_bound = col_mean + deviations * col_std
        outliers = frame[(frame[col] < lower_bound) | (
            frame[col] > upper_bound)][[label_column_name, col]]
        
        # if there are any outliers, add them to the dictionary
        if not outliers.empty:
            outlier_info[col] = list(outliers.itertuples(index=False, name=None))
            
    return outlier_info

print("Outliers found in our data:")
find_outliers(world_data, 4.0, 'Country')

Outliers found in our data:


{}

Once we have found outliers in our data, we have the hard decision of what to do with the outliers:
 - Are the outliers errors in the original data that we should discuss with the original data authors?
 - Are they a symptom of incorrect data parsing earlier in our pipeline?
 - Should they be replaced with a more typical value?
 - Should they be removed entirely?
 - Are they real and should just be kept in the data without modification.

In our case, we will keep our outliers intact.
However, use this as an opportunity to look for bugs in you number parsing code from Part 1.
Incorrectly parsed numbers can easily show up as outliers.

### Data Imputation

Besides outliers, another form of anomalous data we may see are *missing values*.
Missing values may be represented in many different ways depending on the dataset:
numpy.nan, -1, null, None, "MISSING", etc.

Sometimes what to do with missing data is obvious given the dataset.
For example, if a new country was founded in 2022 then the value in its "Unemployment rate 2019" should be missing.
Neither a 0.0 nor 1.0 would make sense, the value just does not exist.

However in more complex situations, deciding what to do with missing data can be very difficult.
One potential way to deal with missing data is through [data imputation](https://en.wikipedia.org/wiki/Imputation_(statistics)).
"Imputing" data is when we replace missing values with some derived value.
We are essentially "filling in the gaps" in our data with values that make sense.
There are many methods of choosing which values to impute, some of which are:
 - Using the **mean** value of the column (using non-empty values).
 - Using other values to create a machine learning model (often a regressor) to **predict** what the value should be.
 - Making a **random** guess (which can sometimes work better than you may expect).
 - Using knowledge of the domain to build a probabilistic model and then using that model to make a **educated random** guess.

There is no fixed rule or theorem that will tell you what the best data imputation method is for each dataset or column.
It will usually come down to the expertise of a data scientists (that's you!)
to bridge the gap between the domain data and statistics and figure our the best way to impute missing data (or whether to leave it be).
Always be careful when imputing values,
since it may bias your data.

In our data, the most sparsely populated columns (and candidates for imputation)
are "Literacy" and the three "Unemployment" columns.

The "Literacy" column could be a very interesting column to impute,
because there are many [studies on different predictors of literacy in adults](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0243763).
We may be able to use the other columns in our dataset (like "Electricity access", "Internet access", "Improved Sanitation", "Improved Water")
to predict literacy rate.
Unfortunately, building regressors is outside the scope of this assignment.

Still, the "Unemployment" columns also provides an interesting data imputation opportunity for us.
There are three columns that represent employment for three years ("Unemployment rate 2021", "Unemployment rate 2020", "Unemployment rate 2019"),
and each country may have values for any number of those columns.
To smooth out the data and impute missing values,
we can combine these columns into one ("Recent Unemployment") and take the **mean** of all present values.

<h3 style="color: darkorange";>★ Task 2.B</h3>

Complete the function below that takes a frame, a list of column names, and a new column name.
Modify the frame to merge the provided columns into a single column with the provided name.
To merge columns, take the mean of any present values.
If no values are present, use `numpy.nan`.
Return the passed-in frame.

In [41]:
'''
Complete the function below that takes a frame, a list of column names, and a new column name.
Modify the frame to merge the provided columns into a single column with the provided name.
To merge columns, take the mean of any present values.
If no values are present, use `numpy.nan`.
Return the passed-in frame.
'''

def merge_columns(frame, merge_column_names, new_column_name):
    frame[new_column_name] = frame[merge_column_names].mean(axis = 1)
    frame.drop(columns = merge_column_names, inplace = True)
    return frame
    

print("Data with merged unemployment data:")
columns = ['Unemployment rate 2019', 'Unemployment rate 2020', 'Unemployment rate 2021']
merge_columns(world_data, columns, 'Recent Unemployment')

Data with merged unemployment data:


Country Life expectancy Population Literacy Electricity access   
0       Afghanistan             NaN        NaN    0.373               0.99  \
1          Akrotiri             NaN        NaN      NaN                NaN   
2           Albania             NaN        NaN    0.981                  1   
3           Algeria             NaN        NaN    0.814              0.994   
4    American Samoa             NaN        NaN      NaN                NaN   
..              ...             ...        ...      ...                ...   
245  Western Sahara             NaN        NaN      NaN                NaN   
246           Yemen             NaN        NaN    0.701               0.47   
247          Zambia             NaN        NaN    0.867               0.37   
248        Zimbabwe             NaN        NaN    0.865               0.53   
249            none             NaN        NaN      NaN               0.82   

    Internet access Improved Sanitation Improved Water   
0              0.18               0.614          0.765  \
1               NaN                 NaN            NaN   
2              0.72               0.999           0.97   
3              0.63               0.965          0.994   
4              0.31                0.99          0.998   
..              ...                 ...            ...   
245             NaN                 NaN            NaN   
246            0.27               0.591          0.996   
247             0.2               0.517          0.716   
248            0.29               0.642          0.769   
249            0.35               0.883          0.786   

                                    Export commodities Labor Agriculture  ...   
0    gold, grapes, opium, fruits and nuts, insect r...             0.443  ...  \
1                                                  NaN               NaN  ...   
2    leather footwear and parts, crude petroleum, i...             0.414  ...   
3    crude petroleum, natural gas, refined petroleu...             0.108  ...   
4                                          canned tuna               NaN  ...   
..                                                 ...               ...  ...   
245                         phosphates 62% (2012 est.)               0.5  ...   
246  crude petroleum, gold, fish, industrial chemic...               NaN  ...   
247  copper, gold, gemstones, sulfuric acid, raw su...             0.548  ...   
248  gold, tobacco, iron alloys, nickel, diamonds, ...             0.675  ...   
249  fish and fish products, coral/shells, scrap me...             0.009  ...   

    Unemployment rate 2014 Unemployment rate 1998 Unemployment rate 2015   
0                      NaN                    NaN                    NaN  \
1                      NaN                    NaN                    NaN   
2                      NaN                    NaN                    NaN   
3                      NaN                    NaN                    NaN   
4                      NaN                    NaN                    NaN   
..                     ...                    ...                    ...   
245                    NaN                    NaN                    NaN   
246                    NaN                    NaN                    NaN   
247                    NaN                    NaN                    NaN   
248                    NaN                    NaN                    NaN   
249                    NaN                    NaN                    NaN   

    Unemployment rate 2002 Unemployment rate 2018   
0                      NaN                    NaN  \
1                      NaN                    NaN   
2                      NaN                    NaN   
3                      NaN                    NaN   
4                      NaN                    NaN   
..                     ...                    ...   
245                    NaN                    NaN   
246                    NaN                    NaN   
247            

## Part 3: Encoding Data

Let's finally deal with that pesky "Export commodities" column!
It's not a number, but represents a list of items.
It is common to deal with data that is not numeric (arbitrary strings, strings representing single items, strings representing list/sets of items, etc).
To deal with these non-numeric columns, we need to encode them.
That is, we need to convert them from a non-numeric form to a numeric one.

In computer science, there are many encodings and encoding methods.
An encoding that you are probably already familiar with is [ASCII](https://en.wikipedia.org/wiki/ASCII),
which is a way to represent Latin letters as numbers.
Encodings are just agreed upon ways to represent non-numeric values as numbers.

In machine learning and data science, probably the most popular encoding method is [the One-Hot encoding](https://en.wikipedia.org/wiki/One-hot).
In One-Hot encoding, we take all the possible values an item can take and assign each possible value an index.
Presence of that value results in a 1, which absence results in a 0.
Essentially, we are turning a list of items into a series of binary columns which ask "Do you have this value?".

For example, consider the following tables.
In the first one, a student's major is kept as a string (or a list of strings).
To encode it, a column is created for each possible major and a 1 is present if that student has that major
(otherwise a 0 is used).

<center><img src="one-hot.png"/></center>

<h3 style="color: darkorange";>★ Task 3.A</h3>

Complete the function below that takes in a frame and a column name.
The function should modify the frame to add multiple columns represented a one-hot encoding of the specified column.
Assume that the given column contains a comma-separated list of values.
Each value should be made lowercase and have additional whitespace removed from the beginning and end.
The new column names should be named: "`<old column name>: <value>`.
(Like in the students example above.)
Return **a new frame** with the columns of the passed-in frame (minus the specified column) and all the new one-hot columns.

Note that the above semantics will not work perfectly for our world data,
but it will provide a good starting point.

In [98]:
def one_hot(frame, column_name):
    
    if column_name not in frame.columns:
        return frame
    
    new_frame = frame.copy()
    values = set()
    
    # Extract all the possible values from the column
    for row in new_frame[column_name]:
        values.update([value.strip().lower() for value in row.split(",")])
    
    # Add one-hot columns for each possible value
    for value in values:
        new_column_name = f"{column_name}: {value}"
        new_frame[new_column_name] = new_frame[column_name].apply(lambda row:
                                    int(value.strip().lower() in row.lower().split(",")))
    
    # Remove the original column
    new_frame.drop(columns=[column_name], inplace=True)
    
    return new_frame

In [102]:
# Before calling your function, we are going to clean up the data a little bit.
# You are not required to learn what we are doing below (regular expressions),
# but they are very useful and will always be useful to you.
world_data['Export commodities'].replace(r'\s*\(\d{4}(?:\s*est\.)?\)\s*', '',
                                         regex = True, inplace = True)
world_data['Export commodities'].replace(r'\s*\(\d{2}%\)\s*', '',
                                         regex = True, inplace = True)
world_data['Export commodities'].replace(r'\s*\d{2}%\s*', '',
                                         regex = True, inplace = True)

print("One-Hot encoded exports:")
world_data = one_hot(world_data, 'Export commodities')
world_data

One-Hot encoded exports:


Country Life expectancy Population Literacy Electricity access   
0       Afghanistan             NaN        NaN    0.373               0.99  \
1          Akrotiri             NaN        NaN      NaN                NaN   
2           Albania             NaN        NaN    0.981                  1   
3           Algeria             NaN        NaN    0.814              0.994   
4    American Samoa             NaN        NaN      NaN                NaN   
..              ...             ...        ...      ...                ...   
245  Western Sahara             NaN        NaN      NaN                NaN   
246           Yemen             NaN        NaN    0.701               0.47   
247          Zambia             NaN        NaN    0.867               0.37   
248        Zimbabwe             NaN        NaN    0.865               0.53   
249            none             NaN        NaN      NaN               0.82   

    Internet access Improved Sanitation Improved Water Labor Agriculture   
0              0.18               0.614          0.765             0.443  \
1               NaN                 NaN            NaN               NaN   
2              0.72               0.999           0.97             0.414   
3              0.63               0.965          0.994             0.108   
4              0.31                0.99          0.998               NaN   
..              ...                 ...            ...               ...   
245             NaN                 NaN            NaN               0.5   
246            0.27               0.591          0.996               NaN   
247             0.2               0.517          0.716             0.548   
248            0.29               0.642          0.769             0.675   
249            0.35               0.883          0.786             0.009   

    Labor Industry  ...   
0            0.181  ...  \
1              NaN  ...   
2            0.183  ...   
3            0.309  ...   
4            0.155  ...   
..             ...  ...   
245            0.5  ...   
246            NaN  ...   
247          0.099  ...   
248          0.073  ...   
249          0.052  ...   

    Export commodities: crude petroleum, refined petroleum, industrial alcohols, gold, iron   
0                                                False                                       \
1                                                  NaN                                        
2                                                False                                        
3                                                False                                        
4                                                False                                        
..                                                 ...                                        
245                                              False                                        
246                                              False                                        
247                                              False                                        
248                                              False                                        
249                                              False                                        

    Export commodities: broadcasting equipment, telephones, integrated circuits, footwear, furniture   
0                                                False                                                \
1                                                  NaN                                                 
2                                                False                                                 
3                                                False                                                 
4                                                False                                                 
..                                                 ...                  

## Part 4: Joining Data

Data is not always from a single frame/table or source.
You will often need to merge and cross-reference data from other sources and in different frames.
When we talk about merging data from multiple frames or tables,
we usually use terminology that comes from databases and relational algebra.
So we say "table" to refer to the abstract concept of tabular data (data with rows and columns) like a Pandas.DataFrame.
Then, a [join](https://en.wikipedia.org/wiki/Join_(SQL)) is an operation that combines data from the columns of two different tables.
(Note that the same table can actually be used multiple times (called a "self-join"),
and joins can be chained together to include as many tables as desired.)

In this section, we will briefly show the main different types of joins.
The image below shows a graphical representation of the primary join types.
When discussing joins, we will refer to the table on the left side of the join operator as the LHS (left-hand side)
and the table on the right side of the join operator as the RHS (right-hand side).
In the diagram below, $ A $ is always the LHS and $ B $ is always the RHS.

Joins will typically specify some condition that must be meant for two rows to be merged,
referred to as a "join condition".
The most common join condition is equality for a set of columns.
When a collection of columns are checked for equality in a join,
we say we join "on" those columns.
For example, we may say that we join $ A $ and $ B $ on the "ID" column.

<center><img src="joins.png"/></center>

### Cross Join / Cartesian Product / Cross Product

Cross joins (also known an "Cartesian products" or "cross products")
are when every row in the LHS is paired with every row in the RHS.
Therefore, the resulting table will have $ |A| * |B| $ number of rows.

Cross joins can be though of as the basis for all other joins,
since all possible rows are enumerated.

### Inner Join

Inner joins take only the rows from the LHS and RHS where the join conditions are met.
We call an inner join where all shared columns (columns with the same name) between the LHS and RHS are checked for equality a *natural join*.

### Full Outer Join

Full outer joins (also known as "outer joins") ensure that every row from each table is represented.
First, rows where the join condition are met are included.
Then, rows from each table are included where the join condition are not met are included.
For these non-matching rows, any columns that come from the other table are filled in with null or empty values
(depending on the engine that performs the join).

### Left Outer Join

Left outer joins are like full outer joins,
except only the non-matching rows from the LHS are included.

### Right Outer Join

Left outer joins are like full outer joins,
except only the non-matching rows from the RHS are included.

Now that you are an expert in join types, let's get some new data to join!
To augment our data, we can get more information from the [World Health Organization](https://www.who.int/) (WHO).
To tie in with our previous HO, we can get some 2022 statistics for each country on their Covid-19 status.

In [93]:
with open('who_covid_data_2022.json', 'r') as file:
    data = json.load(file)

who_data = pandas.DataFrame.from_dict(data, orient = 'index')
who_data.sort_index(axis = 0, inplace = True)
who_data.head()

Code  Cumulative_Confirmed  Per_100000_Confirmed   
Afghanistan      AF                207969               534.235  \
Albania          AL                333072             11573.841   
Algeria          DZ                271255               618.583   
American Samoa   AS                  8309             15053.354   
Andorra          AD                 47781             61840.419   

                Last_7_Days_Confirmed  Last_7_Days_Confirmed_Per_100000   
Afghanistan                       159                             0.408  \
Albania                           134                             4.656   
Algeria                             1                             0.002   
American Samoa                      0                             0.000   
Andorra                             0                             0.000   

                Confirmed  Cumulative_Deaths  Per_100000_Deaths   
Afghanistan            16               7855             20.178  \
Albania                 8               3596            124.957   
Algeria                 0               6881             15.692   
American Samoa          0                 34             61.598   
Andorra                 0                159            205.785   

                Last_7_Days_Deaths  Last_7_Days_Deaths_Per_100000  Deaths   
Afghanistan                      3                          0.008       0  \
Albania                          0                          0.000       0   
Algeria                          0                          0.000       0   
American Samoa                   0                          0.000       0   
Andorra                          0                          0.000       0   

                Total_Vaccinations  Total_Vaccinations_Per_100   
Afghanistan             12526397.0                      32.178  \
Albania                  3025728.0                     105.100   
Algeria                 15267442.0                      34.817   
American Samoa            111316.0                     201.670   
Andorra                   154367.0                     199.800   

                Total_One_Dose  Total_One_Dose_Per_100  Total_Fully_Vacc   
Afghanistan         11533150.0                  29.627        10825004.0  \
Albania              1342243.0                  47.163         1270016.0   
Algeria              7840131.0                  17.879         6481186.0   
American Samoa         44885.0                  81.318           41423.0   
Andorra                57901.0                  76.009           53485.0   

                Total_Fully_Vacc_Per_100  Total_Boosted  Total_Boosted_Per_100  
Afghanistan                       27.808            NaN                    NaN  
Albania                           44.625       377057.0                 13.249  
Algeria                           14.780       575651.0                  1.313  
American Samoa                    75.046        24160.0                 43.770  
Andorra                           70.211        42981.0                 56.423

To perform joins in Pandas, a useful method is [DataFrame.join()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html).
Of course, Pandas always has multiple ways to do anything, but the `DataFrame.join()` method is a good start.

<h3 style="color: darkorange";>★ Task 4.A</h3>

Complete the function below that takes in two frames and a list of column names.
This function should return a new frame that contains the left outer join between these frames on the specified columns from the LHS.

*Hint: Make sure to read the documentation on DataFrame.join() if you are going to use it.*

In [94]:
def left_join(lhs_frame, rhs_frame, column_names):
    return lhs_frame.join(rhs_frame, on = column_names, how = 'left')

print("World data that includes Covid-19 stats:")
covid_world_data = left_join(world_data, who_data, ['Country'])
covid_world_data

World data that includes Covid-19 stats:


Country Life expectancy Population Literacy Electricity access   
0       Afghanistan             NaN        NaN    0.373               0.99  \
1          Akrotiri             NaN        NaN      NaN                NaN   
2           Albania             NaN        NaN    0.981                  1   
3           Algeria             NaN        NaN    0.814              0.994   
4    American Samoa             NaN        NaN      NaN                NaN   
..              ...             ...        ...      ...                ...   
245  Western Sahara             NaN        NaN      NaN                NaN   
246           Yemen             NaN        NaN    0.701               0.47   
247          Zambia             NaN        NaN    0.867               0.37   
248        Zimbabwe             NaN        NaN    0.865               0.53   
249            none             NaN        NaN      NaN               0.82   

    Internet access Improved Sanitation Improved Water Labor Agriculture   
0              0.18               0.614          0.765             0.443  \
1               NaN                 NaN            NaN               NaN   
2              0.72               0.999           0.97             0.414   
3              0.63               0.965          0.994             0.108   
4              0.31                0.99          0.998               NaN   
..              ...                 ...            ...               ...   
245             NaN                 NaN            NaN               0.5   
246            0.27               0.591          0.996               NaN   
247             0.2               0.517          0.716             0.548   
248            0.29               0.642          0.769             0.675   
249            0.35               0.883          0.786             0.009   

    Labor Industry  ... Last_7_Days_Deaths_Per_100000 Deaths   
0            0.181  ...                         0.008    0.0  \
1              NaN  ...                           NaN    NaN   
2            0.183  ...                         0.000    0.0   
3            0.309  ...                         0.000    0.0   
4            0.155  ...                         0.000    0.0   
..             ...  ...                           ...    ...   
245            0.5  ...                           NaN    NaN   
246            NaN  ...                         0.000    0.0   
247          0.099  ...                         0.000    0.0   
248          0.073  ...                         0.000    0.0   
249          0.052  ...                           NaN    NaN   

    Total_Vaccinations Total_Vaccinations_Per_100 Total_One_Dose   
0           12526397.0                     32.178     11533150.0  \
1                  NaN                        NaN            NaN   
2            3025728.0                    105.100      1342243.0   
3           15267442.0                     34.817      7840131.0   
4             111316.0                    201.670        44885.0   
..                 ...                        ...            ...   
245                NaN                        NaN            NaN   
246          1242982.0                      4.167      1006274.0   
247         13581376.0                     73.876     10937674.0   
248         12694853.0                     85.413      6558577.0   
249                NaN                        NaN            NaN   

    Total_One_Dose_Per_100 Total_Fully_Vacc Total_Fully_Vacc_Per_100   
0                   29.627       10825004.0                   27.808  \
1                      NaN              NaN                      NaN   
2                   47.163        1270016.0                   44.625   
3                   17.879        6481186.0                   14.780   
4                   81.318          41423.0                   75.046   
..                     ...              ...                      ...   
245                    NaN              NaN                      NaN   


Now we can see our country data along with Covid-19 data for each country!

In [95]:
covid_world_data.describe()

Export commodities: nan  Cumulative_Confirmed  Per_100000_Confirmed   
count                      0.0          1.940000e+02            194.000000  \
mean                       NaN          2.280567e+06          19563.683567   
std                        NaN          6.476428e+06          19540.472583   
min                        NaN          0.000000e+00              0.000000   
25%                        NaN          2.369825e+04           1787.886750   
50%                        NaN          2.041865e+05          13450.886500   
75%                        NaN          1.293760e+06          31513.233750   
max                        NaN          4.468115e+07          71853.424000   

       Last_7_Days_Confirmed  Last_7_Days_Confirmed_Per_100000      Confirmed   
count             194.000000                        194.000000     194.000000  \
mean             8808.371134                         54.633304     739.355670   
std             70648.118413                        399.645651    7936.837958   
min                 0.000000                          0.000000       0.000000   
25%                 0.000000                          0.000000       0.000000   
50%                 0.000000                          0.000000       0.000000   
75%               303.500000                         12.036500       0.000000   
max            943064.000000                       5438.813000  108281.000000   

       Cumulative_Deaths  Per_100000_Deaths  Last_7_Days_Deaths   
count         194.000000         194.000000          194.000000  \
mean        23756.391753         126.782242           37.850515   
std         74043.277400         131.096094          225.000342   
min             0.000000           0.000000            0.000000   
25%           165.250000          14.328000            0.000000   
50%          2147.500000          88.374000            0.000000   
75%         14062.750000         205.259750            3.000000   
max        695088.000000         662.550000         2857.000000   

       Last_7_Days_Deaths_Per_100000      Deaths  Total_Vaccinations   
count                     194.000000  194.000000        1.910000e+02  \
mean                        0.218227    3.479381        5.840495e+07   
std                         0.672357   31.150181        3.019056e+08   
min                         0.000000    0.000000        1.380000e+02   
25%                         0.000000    0.000000        5.054580e+05   
50%                         0.000000    0.000000        4.523381e+06   
75%                         0.055500    0.000000        2.010046e+07   
max                         5.693000  415.000000        3.465114e+09   

       Total_Vaccinations_Per_100  Total_One_Dose  Total_One_Dose_Per_100   
count                  191.000000    1.920000e+02              192.000000  \
mean                   152.249194    2.428285e+07               62.836734   
std                     85.124399    1.219068e+08               26.471725   
min                      0.256000    0.000000e+00                0.000000   
25%                     80.356500    2.695368e+05               44.057000   
50%                    161.900000    2.789580e+06               67.455000   
75%                    215.910000    9.382555e+06               82.212750   
max                    469.778000    1.307512e+09              163.185000   

       Total_Fully_Vacc  Total_Fully_Vacc_Per_100  Total_Boosted   
count      1.920000e+02                192.000000   1.760000e+02  \
mean       2.242500e+07                 57.710062   1.186132e+07   
std        1.164194e+08                 26.331122   6.412856e+07   
min        0.000000e+00                  0.000000   0.000000e+00   
25%        2.203640e+05                 38.151250   4.532600e+04   
50%        2.504098e+06                 63.219000   6.256960e+05   
75%        8.168116e+06                 78.394500   4.546497e+06   
max        1.277357e+09                163.185000   8.048508e+08   

       Total_B

Note that we will likely have rows from our world data that did not match up with the WHO data.
To see these, we can just look for empty values in WHO columns (since a left join would put empty values there).

In [96]:
covid_world_data[covid_world_data['Total_Vaccinations'].isnull()]

Country Life expectancy   
1                                             Akrotiri             NaN  \
8                                           Antarctica             NaN   
13                         Ashmore and Cartier Islands             NaN   
18   Baker Island, Howland Island, Jarvis Island, J...             NaN   
27                                             Bolivia             NaN   
30                                       Bouvet Island             NaN   
33                                              Brunei             NaN   
36                                               Burma             NaN   
46                                    Christmas Island             NaN   
47                                   Clipperton Island             NaN   
48                             Cocos (Keeling) Islands             NaN   
51                                 Congo (Brazzaville)             NaN   
53                                   Coral Sea Islands             NaN   
55                                       Cote d'Ivoire             NaN   
58                                             Curacao             NaN   
61                                                 DRC             NaN   
63                                            Dhekelia             NaN   
70                                             Eritrea             NaN   
74                   Falkland Islands (Islas Malvinas)            77.9   
80                 French Southern and Antarctic Lands             NaN   
82                                          Gaza Strip             NaN   
97                   Heard Island and McDonald Islands             NaN   
98                             Holy See (Vatican City)             NaN   
100                                          Hong Kong             NaN   
105                                               Iran             NaN   
112                                          Jan Mayen             NaN   
119                                             Kosovo             NaN   
122                                               Laos             NaN   
131                                              Macau             NaN   
142                                            Moldova             NaN   
151                                     Navassa Island             NaN   
160                                     Norfolk Island             NaN   
161                                        North Korea             NaN   
163                           Northern Mariana Islands             NaN   
170                                    Paracel Islands             NaN   
180                                             Russia             NaN   
182                                   Saint Barthelemy             NaN   
185                                       Saint Martin             NaN   
186                          Saint Pierre and Miquelon             NaN   
203           South Georgia and South Sandwich Islands             NaN   
204                                        South Korea             NaN   
207                                    Spratly Islands             NaN   
211  Svalbard (sometimes referred to as Spitsbergen...             NaN   
214                                              Syria             NaN   
215                                             Taiwan             NaN   
217                                           Tanzania             NaN   
219                                        The Bahamas             NaN   
220                                      The Dominican             NaN   
221                                         The Gambia             NaN   
228                                             Turkey             NaN   
234                                     United Kingdom             NaN   
235                                      United States             NaN   
239                                          Venezuela             NaN   
240                                            Viet

Looking at these results, we can understand why the WHO may not have data on some of them.
Some are not technically countries (like "Antarctica"),
some have ambiguous national status (like "Taiwan"),
and some are just using different but valid names (like "United States" vs "United States of America").

As a data scientist, think about ways that you could more efficiently join together these datasets.

Congratulations!
You now have a grasp on the basics of data cleaning and munging (the soul of a data scientist).